view.php

```diff
@@ -24,62 +24,39 @@
  */
 
 require('../../config.php');
+require('./classes/datalib.php');
 
 global $PAGE;
 global $DB;
 
-$id = optional_param('id', 0, PARAM_INT);// Course module ID.
-$observationid = optional_param('o', 0, PARAM_INT);// Observation instance ID.
+$id = optional_param('id', null, PARAM_INT);// Course module ID.
+$observationid = optional_param('o', null, PARAM_INT);// Observation instance ID.
 
-// Can access directly from observation ID or from course module ID.
-if ($observationid) {
-    // Access directly via observation ID.
-    if (!$cm = get_coursemodule_from_instance('observation', $observationid)) {
-        throw new moodle_exception('invalidcoursemodule');
-    }
-
-    // Get the course from the course module (or error).
-    if (!$course = $DB->get_record('course', array('id' => $cm->course))) {
-        throw new moodle_exception('coursemisconf');
-    }
-} else {
-    // Access indirectly via course module ID.
-
-    // Get the course module object from the ID (or error).
-    if (!$cm = get_coursemodule_from_id('observation', $id)) {
-        throw new moodle_exception('invalidcoursemodule');
-    }
-    // Get the course from the course module (or error).
-    if (!$course = $DB->get_record('course', array('id' => $cm->course))) {
-        throw new moodle_exception('coursemisconf');
-    }
+// Needs at least one of the two optional parameters.
+if ($id === null && $observationid === null) {
```

Comment:
if you empty($id) it will be a little more robust, it checks for null, empty string etc. see https://www.php.net/manual/en/function.empty.php